File: String/baekjoon/Kmp.cpp
//https://www.acmicpc.net/problem/1786
//baekjoon-1786

#include <iostream>
#include <string>
#include <vector>
using namespace std;

vector<int> getPi(string P) {
	int j = 0, n = P.length();
	vector<int> pi(n,0);
	
	for(int i=1; i<n; i++) {
		while(j>0 && P[i]!=P[j]){
			j=pi[j-1];
		}
		
		if(P[i]==P[j]){
			pi[i]=++j;
		}
	}

	return pi;
}

vector<int> kmp(string T, string P){
	vector<int> positions, pi = getPi(P);
	
	int n = T.length();
	int m = P.length();
	int j = 0;

	for(int i=0; i<n; i++){
		while(j>0 && T[i]!=P[j]){
			j = pi[j-1];	
		}

		if (T[i]==P[j]) {
			if (j>=m-1) {
				positions.push_back(i-m+1);
				j = pi[j];
			} else {
				j++;
			}
		}
	}

	return positions;
	
}

int main() {
	string T, P;
	getline(cin,T);
	getline(cin,P);

	vector<int> ret = kmp(T,P);

	cout << ret.size() << endl;
	for(int i=0; i<ret.size(); i++) {
		cout << ret[i]+1 << " ";
	}
	return 0;
}

File: String/baekjoon/Kmp.go
package main

import "fmt"

func getPi(p string) []int {
	n := len(p)
	pi := make([]int, n)
	j := 0

	for i := 1; i < n; i++ {
		for j > 0 && p[i] != p[j] {
			j = pi[j-1]
		}
		if p[i] == p[j] {
			j += 1
			pi[i] = j
		}
	}

	return pi
}

func kmp(t, p string) (ret []int) {
	n := len(t)
	pi := getPi(p)
	j := 0

	for i := 0; i < n; i++ {
		for j > 0 && t[i] != p[j] {
			j = pi[j-1]
		}

		if t[i] == p[j] {
			if j >= len(p)-1 {
				ret = append(ret, i-len(p)+1)
			} else {
				j++
			}
		}
	}

	return
}

func main() {
	var T, P string
	fmt.Scanln(&T)
	fmt.Scanln(&P)

	ret := kmp(T, P)

	fmt.Println(len(ret))
	for _, v := range ret {
		fmt.Printf("%d ", v+1)
	}
	fmt.Println()
}
